File: src-tauri/src/commands/usage.rs
//! 使用统计相关命令

use crate::error::AppError;
use crate::services::usage_pricing::{self, ModelPricingInfo};
use crate::services::usage_stats::*;
use crate::store::AppState;
use tauri::State;

/// 获取使用量汇总
#[tauri::command]
pub fn get_usage_summary(
    state: State<'_, AppState>,
    start_date: Option<i64>,
    end_date: Option<i64>,
    app_type: Option<String>,
    provider_name: Option<String>,
    model: Option<String>,
) -> Result<UsageSummary, AppError> {
    state.db.get_usage_summary(
        start_date,
        end_date,
        app_type.as_deref(),
        provider_name.as_deref(),
        model.as_deref(),
    )
}

/// 获取按 app_type 拆分的使用量汇总
#[tauri::command]
pub fn get_usage_summary_by_app(
    state: State<'_, AppState>,
    start_date: Option<i64>,
    end_date: Option<i64>,
    provider_name: Option<String>,
    model: Option<String>,
) -> Result<Vec<UsageSummaryByApp>, AppError> {
    state.db.get_usage_summary_by_app(
        start_date,
        end_date,
        provider_name.as_deref(),
        model.as_deref(),
    )
}

/// 获取每日趋势
#[tauri::command]
pub fn get_usage_trends(
    state: State<'_, AppState>,
    start_date: Option<i64>,
    end_date: Option<i64>,
    app_type: Option<String>,
    provider_name: Option<String>,
    model: Option<String>,
) -> Result<Vec<DailyStats>, AppError> {
    state.db.get_daily_trends(
        start_date,
        end_date,
        app_type.as_deref(),
        provider_name.as_deref(),
        model.as_deref(),
    )
}

/// 获取 Provider 统计
#[tauri::command]
pub fn get_provider_stats(
    state: State<'_, AppState>,
    start_date: Option<i64>,
    end_date: Option<i64>,
    app_type: Option<String>,
    provider_name: Option<String>,
    model: Option<String>,
) -> Result<Vec<ProviderStats>, AppError> {
    state.db.get_provider_stats(
        start_date,
        end_date,
        app_type.as_deref(),
        provider_name.as_deref(),
        model.as_deref(),
    )
}

/// 获取模型统计
#[tauri::command]
pub fn get_model_stats(
    state: State<'_, AppState>,
    start_date: Option<i64>,
    end_date: Option<i64>,
    app_type: Option<String>,
    provider_name: Option<String>,
    model: Option<String>,
) -> Result<Vec<ModelStats>, AppError> {
    state.db.get_model_stats(
        start_date,
        end_date,
        app_type.as_deref(),
        provider_name.as_deref(),
        model.as_deref(),
    )
}

/// 获取请求日志列表
#[tauri::command]
pub fn get_request_logs(
    state: State<'_, AppState>,
    filters: LogFilters,
    page: u32,
    page_size: u32,
) -> Result<PaginatedLogs, AppError> {
    state.db.get_request_logs(&filters, page, page_size)
}

/// 获取单个请求详情
#[tauri::command]
pub fn get_request_detail(
    state: State<'_, AppState>,
    request_id: String,
) -> Result<Option<RequestLogDetail>, AppError> {
    state.db.get_request_detail(&request_id)
}

/// 获取模型定价列表
#[tauri::command]
pub fn get_model_pricing(state: State<'_, AppState>) -> Result<Vec<ModelPricingInfo>, AppError> {
    usage_pricing::get_model_pricing(&state.db)
}

/// 更新模型定价
#[tauri::command]
pub fn update_model_pricing(
    state: State<'_, AppState>,
    model_id: String,
    display_name: String,
    input_cost: String,
    output_cost: String,
    cache_read_cost: String,
    cache_creation_cost: String,
) -> Result<(), AppError> {
    usage_pricing::update_model_pricing(
        &state.db,
        model_id,
        display_name,
        input_cost,
        output_cost,
        cache_read_cost,
        cache_creation_cost,
    )
}

/// 检查 Provider 使用限额
#[tauri::command]
pub fn check_provider_limits(
    state: State<'_, AppState>,
    provider_id: String,
    app_type: String,
) -> Result<crate::services::usage_stats::ProviderLimitStatus, AppError> {
    state.db.check_provider_limits(&provider_id, &app_type)
}

/// 删除模型定价
#[tauri::command]
pub fn delete_model_pricing(state: State<'_, AppState>, model_id: String) -> Result<(), AppError> {
    usage_pricing::delete_model_pricing(&state.db, model_id)
}

/// 手动触发会话日志同步
#[tauri::command]
pub fn sync_session_usage(
    state: State<'_, AppState>,
) -> Result<crate::services::session_usage::SessionSyncResult, AppError> {
    crate::services::session_usage_sync::sync_all_session_usage(&state.db)
}

/// 获取数据来源分布
#[tauri::command]
pub fn get_usage_data_sources(
    state: State<'_, AppState>,
) -> Result<Vec<crate::services::session_usage::DataSourceSummary>, AppError> {
    crate::services::session_usage::get_data_source_breakdown(&state.db)
}

File: src/lib/api/usage.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  UsageSummary,
  UsageSummaryByApp,
  DailyStats,
  ProviderStats,
  ModelStats,
  RequestLog,
  LogFilters,
  ModelPricing,
  ProviderLimitStatus,
  PaginatedLogs,
  SessionSyncResult,
  DataSourceSummary,
} from "@/types/usage";
import type { UsageResult } from "@/types";
import type { AppId } from "./types";
import type { TemplateType } from "@/config/constants";
import { LOCAL_MANAGEMENT_TARGET } from "@/lib/managementTarget";
import { remoteApi, type ManagementTarget } from "./remote";

export const usageApi = {
  // Provider usage script methods
  query: async (
    providerId: string,
    appId: AppId,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<UsageResult> => {
    if (target.type === "remote") {
      return remoteApi.queryProviderUsage(
        target.profile,
        appId,
        providerId,
        target.secret,
      );
    }

    return invoke("queryProviderUsage", { providerId, app: appId });
  },

  testScript: async (
    providerId: string,
    appId: AppId,
    scriptCode: string,
    timeout?: number,
    apiKey?: string,
    baseUrl?: string,
    accessToken?: string,
    userId?: string,
    templateType?: TemplateType,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<UsageResult> => {
    if (target.type === "remote") {
      return remoteApi.testUsageScript(
        target.profile,
        appId,
        providerId,
        {
          scriptCode,
          timeout,
          apiKey,
          baseUrl,
          accessToken,
          userId,
          templateType,
        },
        target.secret,
      );
    }

    return invoke("testUsageScript", {
      providerId,
      app: appId,
      scriptCode,
      timeout,
      apiKey,
      baseUrl,
      accessToken,
      userId,
      templateType,
    });
  },

  // Proxy usage statistics methods
  getUsageSummary: async (
    startDate?: number,
    endDate?: number,
    appType?: string,
    providerName?: string,
    model?: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<UsageSummary> => {
    if (target.type === "remote") {
      return remoteApi.getUsageSummary(
        target.profile,
        startDate,
        endDate,
        appType,
        providerName,
        model,
        target.secret,
      );
    }
    return invoke("get_usage_summary", {
      startDate,
      endDate,
      appType,
      providerName,
      model,
    });
  },

  getUsageSummaryByApp: async (
    startDate?: number,
    endDate?: number,
    providerName?: string,
    model?: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<UsageSummaryByApp[]> => {
    if (target.type === "remote") {
      return remoteApi.getUsageSummaryByApp(
        target.profile,
        startDate,
        endDate,
        providerName,
        model,
        target.secret,
      );
    }
    return invoke("get_usage_summary_by_app", {
      startDate,
      endDate,
      providerName,
      model,
    });
  },

  getUsageTrends: async (
    startDate?: number,
    endDate?: number,
    appType?: string,
    providerName?: string,
    model?: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<DailyStats[]> => {
    if (target.type === "remote") {
      return remoteApi.getUsageTrends(
        target.profile,
        startDate,
        endDate,
        appType,
        providerName,
        model,
        target.secret,
      );
    }
    return invoke("get_usage_trends", {
      startDate,
      endDate,
      appType,
      providerName,
      model,
    });
  },

  getProviderStats: async (
    startDate?: number,
    endDate?: number,
    appType?: string,
    providerName?: string,
    model?: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<ProviderStats[]> => {
    if (target.type === "remote") {
      return remoteApi.getProviderStats(
        target.profile,
        startDate,
        endDate,
        appType,
        providerName,
        model,
        target.secret,
      );
    }
    return invoke("get_provider_stats", {
      startDate,
      endDate,
      appType,
      providerName,
      model,
    });
  },

  getModelStats: async (
    startDate?: number,
    endDate?: number,
    appType?: string,
    providerName?: string,
    model?: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<ModelStats[]> => {
    if (target.type === "remote") {
      return remoteApi.getModelStats(
        target.profile,
        startDate,
        endDate,
        appType,
        providerName,
        model,
        target.secret,
      );
    }
    return invoke("get_model_stats", {
      startDate,
      endDate,
      appType,
      providerName,
      model,
    });
  },

  getRequestLogs: async (
    filters: LogFilters,
    page: number = 0,
    pageSize: number = 20,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<PaginatedLogs> => {
    if (target.type === "remote") {
      return remoteApi.getRequestLogs(
        target.profile,
        filters,
        page,
        pageSize,
        target.secret,
      );
    }
    return invoke("get_request_logs", {
      filters,
      page,
      pageSize,
    });
  },

  getRequestDetail: async (
    requestId: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<RequestLog | null> => {
    if (target.type === "remote") {
      return remoteApi.getRequestDetail(
        target.profile,
        requestId,
        target.secret,
      );
    }
    return invoke("get_request_detail", { requestId });
  },

  getModelPricing: async (
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<ModelPricing[]> => {
    if (target.type === "remote") {
      return remoteApi.getModelPricing(target.profile, target.secret);
    }
    return invoke("get_model_pricing");
  },

  updateModelPricing: async (
    modelId: string,
    displayName: string,
    inputCost: string,
    outputCost: string,
    cacheReadCost: string,
    cacheCreationCost: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<void> => {
    if (target.type === "remote") {
      return remoteApi.updateModelPricing(
        target.profile,
        modelId,
        displayName,
        inputCost,
        outputCost,
        cacheReadCost,
        cacheCreationCost,
        target.secret,
      );
    }
    return invoke("update_model_pricing", {
      modelId,
      displayName,
      inputCost,
      outputCost,
      cacheReadCost,
      cacheCreationCost,
    });
  },

  deleteModelPricing: async (
    modelId: string,
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<void> => {
    if (target.type === "remote") {
      return remoteApi.deleteModelPricing(
        target.profile,
        modelId,
        target.secret,
      );
    }
    return invoke("delete_model_pricing", { modelId });
  },

  checkProviderLimits: async (
    providerId: string,
    appType: string,
  ): Promise<ProviderLimitStatus> => {
    return invoke("check_provider_limits", { providerId, appType });
  },

  // Session usage sync
  syncSessionUsage: async (
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<SessionSyncResult> => {
    if (target.type === "remote") {
      return remoteApi.syncSessionUsage(target.profile, target.secret);
    }
    return invoke("sync_session_usage");
  },

  rebuildCodexUsage: async (
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<SessionSyncResult> => {
    if (target.type === "remote") {
      return remoteApi.rebuildCodexUsage(target.profile, target.secret);
    }
    return invoke("rebuild_codex_usage");
  },

  getDataSourceBreakdown: async (
    target: ManagementTarget = LOCAL_MANAGEMENT_TARGET,
  ): Promise<DataSourceSummary[]> => {
    if (target.type === "remote") {
      return remoteApi.getDataSourceBreakdown(target.profile, target.secret);
    }
    return invoke("get_usage_data_sources");
  },
};
